packages/ldap-auth-backend/src/ldap.ts

```diff
@@ -0,0 +1,77 @@
+import type { LDAPUser } from './types';
+
+import ldap from 'ldapjs';
+import { dn } from 'ldap-escape';
+import { authenticate, AuthenticationOptions } from 'ldap-authentication';
+
+import { AUTH_USER_NOT_FOUND, LDAP_CONNECT_FAIL } from './errors';
+
+async function _verifyUserExistsNoAdmin(
+    searchString: string,
+    ldapOpts: ldap.ClientOptions,
+    searchOpts = {}
+): Promise<boolean> {
+    const client: ldap.Client = await new Promise((resolve, reject) => {
+        ldapOpts.connectTimeout = ldapOpts.connectTimeout || 5000;
+        const client = ldap.createClient(ldapOpts);
+
+        client.on('connect', () => resolve(client));
+        client.on('timeout', reject);
+        client.on('connectTimeout', reject);
+        client.on('error', reject);
+        client.on('connectError', reject);
+    });
+    if (client instanceof Error)
+        throw new Error(`${LDAP_CONNECT_FAIL} ${client.message}`);
+    return new Promise((resolve, reject) => {
+        client.search(searchString, searchOpts, (error, res) => {
+            if (error) reject(error);
+            let exists = false; // avoids double resolve call if user exists
+            res.on('searchEntry', () => (exists = true));
+            res.on('error', () => client.unbind());
+            res.on('end', () => {
+                resolve(exists);
+                client.unbind();
+            });
+        });
+    });
+}
+
+export const defaultCheckUserExists = async (
+    ldapAuthOptions: AuthenticationOptions,
+    searchFunction = authenticate
+) => {
+    // This fallback is for clients with no need for an admin to list users
+    // I'll remove this if we can add this option to the auth library.
+    if (!ldapAuthOptions.adminDn || !ldapAuthOptions.adminPassword) {
```

Comment:
But here, why do you use `ldapjs` and not `ldap-authentication`? Because if you do not pass the `adminDn` and `adminPassword` do you receive an error?

Reply:
Because `ldap-authentication` forces us to use `adminDn` to check for user existence, I've opened an issue to them but up until this get fixed by them we should provide a way to list users to anyone who doesn't need credential to do so, eg: dev enviroments

Reply:
okay! thank you!